Vizzuality/widget-editor
# Improve the advanced editor and handling advanced widgets

This PR improves the advanced editor and the handling of advanced widgets.

<p align="center">
<img width="1419" alt="Pie chart on the left, advanced editor on the right with some validation error" src="https://user-images.githubusercontent.com/6073968/90020325-a1c65400-dca7-11ea-969e-846fd0095c8c.png">
</p>

In addition, this PR comes with a new wiki page: [Widgets' specification](https://github.com/Vizzuality/widget-editor/wiki/Widgets'-specification).

## Testing instructions

3 general cases are to be tested:
1. Editor instantiated with a dataset only
2. Editor instantiated with simple widget
3. Editor instantiated with advanced widget

For each of these cases, you can use this dataset: `852f2275-91a8-4500-9f10-89880dc53f22`.
For case 2, you can use this widget: `b867b473-d4f9-4105-b87c-df76caaaf358`.
For case 3, you can use this one: `c1d7cb83-5f50-43eb-8ea0-6f7ecd4e1b69`.

1. Instantiate the editor with the dataset (and eventually the widget)
2. Click “Get Editor state”

The payload must be empty if you didn't pass any widget ID, or it must contain the widget's `widgetConfig`.

3. If instantiated without a widget, select a category and value to create a bar chart
4. Click “Get Editor state”

The payload must contain the widget.

5. Click the “Advanced” tab and, if you're not editing an advanced widget, click the “Switch to the advanced mode” button
6. Click “Get Editor state”

The payload must not contain `paramsConfig`.

7. In the “Vega JSON” editor, find the `config` object and edit the fill color of the `rect` object (for bar charts only, edit another part of the `config` otherwise)

After a short moment, the visualisation must update and the bars (or slices, lines, etc.) should be of the colour you selected.

8. Click “Get Editor state”

The payload must contain the same `config` you edited.

9. Click the “click here” link to go back to the interactive mode (if you're not editing an advanced widget)

The visualisation must revert to its original colour.

10. Click “Get Editor state”

The payload must contain a `paramsConfig` object and the `config` object must be restored to its original value.

In addition to these tests, you may also want to test the validation of the advanced editor's content. To do this, check `src/applications/widget-editor/src/components/advanced-editor/helpers.js`.

## Pivotal Tracker

This PR is tracked by these 3 tasks: [task 1](https://www.pivotaltracker.com/story/show/172834856), [task 2](https://www.pivotaltracker.com/story/show/173901209) and [task 3](https://www.pivotaltracker.com/story/show/172822355).
